### PR TITLE
Agr_Pushpa_Edges_Hotel

### PR DESCRIPTION
**Pull request by pushpaagr**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I had originally started with making 3 classes, room, reservation, and booking employee, however ended up changing booking employee to booking system to better reflect the purpose of the class. I then changed my design to then remove the booking system class and put all of my classes into an module called hotel, and replace booking system with only a module hotel since I was not really creating new instances of hotel, and I was really only making new instances of room and reservation. However I decided that will not be a good idea when my module hotel would have to be used by another hotel, or another person in another location(and instance would have to be created), so I changed it back to having a class there which is now named hotel which is inside of the module hotel as well.  
Describe a concept that you gained more clarity on as you worked on this assignment. | I gained clarity on being able to use class methods better when working with multiple classes. I learned that inorder to call instance variable of a class on its instance it has be an instance of that class. For example if I wanted to use a method/instance variable in the Reservation class, I would have to use it on an instance of the reservation class. 
Describe a _nominal test_ that you wrote for this assignment. | I wrote a test that would only return available rooms after reservations had been made.
Describe an _edge case test_ that you wrote for this assignment. | I checked to see what happens when I have made over 20 reservations and all the rooms have been booked to see what my test does, and it will assign a reservation to a nil room, so I raised an argument for that case.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I did okay with writing pseudocode first, mostly did this alot during the first couple days of the project, however I noticed that whenever I became stuck going back to pseudocode helped. I didn't do so great with writing my tests first then writing code. I mostly wrote my code first then wrote the test for it.

### CODE REVIEW

File: lib/hotel.rb [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    end
+
+
+    #returns matched reservations for a specific date
+    def reservations_for_date(date)
+      matches = []
+      @reservations.each do |reservation|
+        if reservation.start_date <= date && reservation.end_date > date
+          matches << reservation
+        end
```
Comment (droberts-sea) [inline lib/hotel.rb:36]: This is an example of tight coupling, because the `Hotel` class needs to know about the details of how a `Reservation` stores its dates. Instead you might write a method like `Reservation#contains?(date)` to encapsulate this logic. Then if the internal logic of a `Reservation` changed, this code would not need to change.

File: lib/hotel.rb [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    end
+
+
+
+    def reserved_rooms_for_dates(start_date, end_date)
+      matched_date_range = []
+      @reservations.each do |reservation|
+        if reservation.start_date.between?(start_date, end_date)
+          matched_date_range << reservation.room_num
+        elsif reservation.end_date.between?(start_date, end_date)
```
Comment (droberts-sea) [inline lib/hotel.rb]: Good work on this tricky logic.

This is another example of tight coupling, which could be replaced by a method like `Reservation#overlap?(date_range)`.

File: lib/hotel.rb [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        end
+      end
+      return matched_date_range
+    end
+
+
+    def nonreserved_rooms_fordates(start_date, end_date)
+      matched_rooms_range = reserved_rooms_for_dates(start_date, end_date)
+      all_room_numbers = get_all_numbers
+      avail_rooms = all_room_numbers - matched_rooms_range
```
Comment (droberts-sea) [inline lib/hotel.rb:61]: This is an excellent example of putting multiple helper methods together to solve a tricky problem - good work.

File: lib/reservation.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+
+    def valid_dates?(start_date, end_date)
+      return start_date < end_date
+    end
+
+    #Date.parse format: Date.parse("10/5/2018"), 10 is day, 5 is month
+    def argument_error_dates(start_date, end_date)
+      if !valid_dates?(@start_date, @end_date)
+        raise ArgumentError, "start date has to before end date start date: #{@start_date}"
```
Comment (droberts-sea) [inline lib/reservation.rb:31]: I like the idea of checking for a valid date range in a separate method, but I think you may have gone a little far in by also splitting out the `valid_dates?` method.

File: lib/reservation.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      end
+    end
+
+    def reservation_cost
+      total = 0
+      total = ((@end_date - @start_date) - 1) * PRICE
+      return total
```
Comment (droberts-sea) [inline lib/reservation.rb:37]: Since this method is already attached to class `Reservation`, you can omit `reservation_` from the name and just call it `cost`.

File: lib/room.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+module Hotel
+
+  class Room
+
+    attr_reader :room_num
+
+    def initialize(room_num)
+      @room_num = room_num
+
+    end
```
Comment (droberts-sea) [inline lib/room.rb:12]: Do you really need a whole class for a `Room`? What responsibilities does it have that you wouldn't get by using integers instead?

File: lib/hotel.rb [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+          matched_date_range << reservation.room_num
+        end
+      end
+      return matched_date_range
+    end
+
+
+    def nonreserved_rooms_fordates(start_date, end_date)
+      matched_rooms_range = reserved_rooms_for_dates(start_date, end_date)
+      all_room_numbers = get_all_numbers
```
Comment (droberts-sea) [inline lib/hotel.rb:60]: Your method names don't match - sometimes you use `for_dates` and sometimes `fordates`.

File: lib/hotel.rb [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    #assings room from the available room's array
+    def get_available_room(start_date, end_date)
+      avail_rooms = nonreserved_rooms_fordates(start_date, end_date)
+      return avail_rooms.first
+    end
+
+    #assigns a reservation given a vaild date, and also calls the get_available_room method which will then return a room number. @reservation is storing all the reservations
+    def assigns_a_reservation(start_date, end_date)
+
+      room_num = get_available_room(start_date, end_date)
```
Comment (droberts-sea) [inline lib/hotel.rb]: We typically name methods as commands rather than descriptions, in which case this would be `assign_reservation`. An even more descriptive name might be something like `build_reservation`.

File: lib/hotel.rb [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    end
+
+
+    #wave 2 copy of list of room, iterate though list of reservation---care
+
+    #returns an array of numbers from a collection of room instances (room_num)
+    def get_all_numbers
+      room_nums = []
+      @rooms.each do |room|
+        room_nums << room.room_num
```
Comment (droberts-sea) [inline lib/hotel.rb:25]: I'm not sure I'm convinced by the method name `get_all_numbers`. I think something like `build_room_list` would do a better job of describing what's going on.

File: lib/reservation.rb (hunk 4 above)
Comment (droberts-sea) [reply lib/reservation.rb:31]: I'm also not convinced by this method name - something like `check_dates` feels better to me.

File: spec/hotel_spec.rb [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+  before do
+    @hotel = Hotel::Hotel.new
+    @start_date = Date.parse("1/9/2018")
+    @end_date = Date.parse("5/9/2018")
+  end
+
+  it 'returns an instance of a room' do
+    room = Hotel::Room.new(1)
+    expect(room).must_be_kind_of Hotel::Room
+  end
```
Comment (droberts-sea) [inline spec/hotel_spec.rb:16]: Why are you testing the `Room` class inside the specs for `Hotel`?

File: spec/hotel_spec.rb [hunk 11]
```diff
@@ -0,0 +1,154 @@
+require_relative 'spec_helper'
+require 'pry'
+
+describe 'Module Hotel' do
+  before do
+    @hotel = Hotel::Hotel.new
+    @start_date = Date.parse("1/9/2018")
```
Comment (droberts-sea) [inline spec/hotel_spec.rb:7]: I would like to see better organization for these tests. My typical strategy is to have a big `describe` for the class, and then a smaller `describe` nested inside for each method.

File: spec/hotel_spec.rb [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    @start_date1 = Date.parse("4/9/2018")
+    @end_date1 = Date.parse("6/9/2018")
+
+    reservation1 = @hotel.assigns_a_reservation(@start_date, @end_date)
+
+    reservation2 = @hotel.assigns_a_reservation(@start_date1, @end_date1)
+
+    expect(@hotel.reservations).must_be_kind_of Array
+    expect(reservation1).must_be_kind_of Hotel::Reservation
+    expect(reservation2).must_be_kind_of Hotel::Reservation
```
Comment (droberts-sea) [inline spec/hotel_spec.rb:50]: These last two assertions don't have anything to do with testing the `Hotel#reservations` method, and probably shouldn't be part of this test. If you wanted to check for inclusion you could do something like:

```ruby
expect(@hotel.reservations).must_include reservation1
expect(@hotel.reservations).must_include reservation2

File: spec/hotel_spec.rb [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+
+  it "returns list of reserved room_nums for a specific
+  DATE RANGE" do
+
+  reservation1 = @hotel.assigns_a_reservation(@start_date, @end_date)
+
+  reservations_fordates = @hotel.reserved_rooms_for_dates(@start_date, @end_date)
+
+  expect(reservations_fordates).must_be_kind_of Array
+  expect(reservation1).must_be_kind_of Hotel::Reservation
```
Comment (droberts-sea) [inline spec/hotel_spec.rb:83]: `reservations_fordates` isn't a great name for this variable, since it really contains a list of room numbers.

File: spec/hotel_spec.rb [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+  expect(avail_rooms).must_be_kind_of Array
+  expect(reservation1).must_be_kind_of Hotel::Reservation
+  expect(avail_rooms).must_equal [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
+
+end
+
+it "reserves an available room for a given date range" do
+
+  22.times do |i|
+    @hotel.assigns_a_reservation((@start_date - i), (@end_date - i))
```
Comment (droberts-sea) [inline spec/hotel_spec.rb:102]: What is this test actually testing? Looping through all these dates doesn't make much sense to me.

File: spec/hotel_spec.rb [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+  expect(@hotel.reservations.length).must_equal 23
+  expect(@hotel.reservations.class).must_equal Array
+
+
+end
+
+it "Raises an ArgumentError when all the rooms have been reserved out and there are no longer any more rooms available" do
+  proc {
+    22.times do |i|
+      @hotel.assigns_a_reservation((@start_date - i), (@end_date + i))
```
Comment (droberts-sea) [inline spec/hotel_spec.rb:123]: I like the idea of this test. However, I don't understand why you're changing the start and end dates. I would also probably be a little more explicit about when I want reserving a room to fail.

```ruby
it "Raises an error if no rooms are available" do
  # Arrange: reserve all the rooms for these dates
  @hotel.rooms.length.times do
    @hotel.assigns_a_reservation(@start_date, @end_date)
  end

  proc {
    # Act: try to reserve another room
    @hotel.assigns_a_reservation(@start_date, @end_date)
  # Assert
  }.must_raise
end
```

File: spec/reservation_spec.rb [hunk 16]
```diff
@@ -0,0 +1,18 @@
+require_relative 'spec_helper'
+
+
+#date structure enter: day, month, year for Date.Parse format
+
+describe 'Reservation Class' do
+  before do
+    @hotel = Hotel::Hotel.new
+    @start_date = Date.parse("5/9/2018")
```
Comment (droberts-sea) [inline spec/reservation_spec.rb:8]: Instead of going through an instance of `Hotel` in these tests, you should be creating instances of the `Reservation` class directly with `Reservation.new`.

File: spec/reservation_spec.rb [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+    @end_date = Date.parse("10/9/2018")
+  end
+
+  it "calculates the total cost per reservation" do
+    reservation = @hotel.assigns_a_reservation(@start_date, @end_date)
+    expect(reservation.reservation_cost).must_equal 800
+  end
```
Comment (droberts-sea) [inline spec/reservation_spec.rb:44]: What happens if you create a reservation with an invalid date range?

File: spec/hotel_spec.rb [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+    room_num = @hotel.get_available_room(@start_date, @end_date)
+    expect(room_num).must_equal 1
+  end
+
+  #Date.parse format: Date.parse("10/5/2018"), 10 is day, 5 is month
+
+  it "raises an argument error for invalid dates" do
+    #proc catches the error when .assigns_a_reservation is invoked
+    proc {
+      @hotel.assigns_a_reservation(@end_date, @start_date)
```
Comment (droberts-sea) [inline spec/hotel_spec.rb]: This functionality lives in `Reservation` not `Hotel`, so I would test it there.

File: spec/hotel_spec.rb [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+    expect(@hotel.reservations).must_be_kind_of Array
+    expect(reservation1).must_be_kind_of Hotel::Reservation
+    expect(reservation2).must_be_kind_of Hotel::Reservation
+
+
+  end
+
+  it "return a reservation for a specific date" do
+
+    match_res = @hotel.assigns_a_reservation(@start_date, @end_date)
```
Comment (droberts-sea) [inline spec/hotel_spec.rb:58]: There's a lot of edge cases around "when is a date part of a reservation" that aren't being covered here.

File: spec/hotel_spec.rb [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+  expect(reservation1).must_be_kind_of Hotel::Reservation
+  expect(reservations_fordates[0]).must_equal 1
+end
+
+
+
+it "gives list of rooms that are available after some of them have been reserved" do
+
+  reservation1 = @hotel.assigns_a_reservation(@start_date, @end_date)
+
```
Comment (droberts-sea) [inline spec/hotel_spec.rb:91]: Similarly, there's a lot of edge cases around room availability that you're missing. See the list that was posted in slack.

**Comment by droberts-sea** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| yes
Answer comprehension questions	| yes
**Design** |
Each class is responsible for a single piece of the program | mostly - I'm not sure `Room` has a whole responsibility
Classes are loosely coupled | see inline
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | needs improvement
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | needs improvement
**Wave 3** |
Create a block of rooms | no
Check if a block has rooms | no 
Reserve a room from a block | no
Test coverage | no
**Fundamentals** | 
Names variables, classes and modules appropriately | some - method names are a little odd sometimes
Understanding of variable scope - local vs instance | yes
Can create complex logical structures utilizing variables | yes
Appropriately uses methods to break down tasks into smaller simpler tasks | yes - good work
Understands the differences between class and instance methods | yes
Appropriately uses iterators and enumerables | yes
Appropriately writes and utilizes classes | yes
Appropriately utilizes modules as a mechanism to encapsulate functionality | yes
**Wrap Up** |
There is a refactors.txt file | no - you'll need this for the break week homework
The file provides a roadmap to future changes | 
**Additional Feedback** | 

This is a good start, but there's a lot of room for improvement here. Let's start with the good:

- The code for waves 1 and 2 works without any obvious bugs
- The flow of your program is easy to follow - you've done a good job of breaking complex work up into small, digestible methods

I see both of these as being key Ruby fundamentals, so it's important that you're doing well here.

Here are areas where you could improve:

- Ability to quickly turn a project write-up into a design
- Naming, especially methods
- Test coverage

All three of the these I've listed above are hard pieces of software engineering, and we don't expect you to be perfect. They're also all about being able to take a step back and reason about your code, not just in small chunks, but as an entire system. Testing is especially important in this regard. We won't add testing to Rails until after the break week, but when we do I would love to sit down with you before the project is due and talk through your test cases.

Keep your focus, and keep up the hard work!

**Comment by pushpaagr** [comment]
Hello Dan,

Did you want to meet today to further discuss testing before break?

Thanks,

Pushpa

File: spec/reservation_spec.rb (hunk 17 above)
Comment (pushpaagr) [reply spec/reservation_spec.rb:44]: Added a test for this scenario

File: spec/hotel_spec.rb (hunk 18 above)
Comment (pushpaagr) [reply spec/hotel_spec.rb]: test removed, and it has already been added to Reservation spec.
